diff -ubBr gc-7.2alpha6.orig/include/private/gcconfig.h gc-7.2alpha6/include/private/gcconfig.h
--- gc-7.2alpha6.orig/include/private/gcconfig.h	Wed Jun  1 00:27:12 2011
+++ gc-7.2alpha6/include/private/gcconfig.h	Sun May 13 17:37:47 2012
@@ -2424,6 +2424,7 @@
 #if ((defined(UNIX_LIKE) && (defined(DARWIN) || defined(HURD) \
                              || defined(OPENBSD) || defined(ARM32) \
                              || defined(MIPS) || defined(AVR32))) \
+     || (defined(FREEBSD) && defined(__FreeBSD__) && __FreeBSD__ < 5) \
      || (defined(LINUX) && (defined(SPARC) || defined(M68K))) \
      || (defined(RTEMS) && defined(I386))) && !defined(NO_GETCONTEXT)
 # define NO_GETCONTEXT
diff -urN gc-7.4.2.orig/os_dep.c gc-7.4.2/os_dep.c
--- gc-7.4.2.orig/os_dep.c	2014-06-03 15:08:02.000000000 +0900
+++ gc-7.4.2/os_dep.c	2020-01-04 16:34:32.296871000 +0900
@@ -3089,6 +3089,10 @@
 #     ifndef SEGV_ACCERR
 #       define SEGV_ACCERR 2
 #     endif
+#     ifndef BUS_PAGE_FAULT
+        /* imported from FreeBSD 11.3 /usr/include/x86/trap.h */
+#       define BUS_PAGE_FAULT T_PAGEFLT
+#     endif
 #     if defined(POWERPC)
 #       define AIM  /* Pretend that we're AIM. */
 #       include <machine/trap.h>
