diff -urN gc-7.4.2.orig/os_dep.c gc-7.4.2/os_dep.c
--- gc-7.4.2.orig/os_dep.c	2014-06-03 15:08:02.000000000 +0900
+++ gc-7.4.2/os_dep.c	2020-01-04 16:34:32.296871000 +0900
@@ -3089,6 +3089,10 @@
 #     ifndef SEGV_ACCERR
 #       define SEGV_ACCERR 2
 #     endif
+#     ifndef BUS_PAGE_FAULT
+        /* imported from FreeBSD 11.3 /usr/include/x86/trap.h */
+#       define BUS_PAGE_FAULT T_PAGEFLT
+#     endif
 #     if defined(POWERPC)
 #       define AIM  /* Pretend that we're AIM. */
 #       include <machine/trap.h>
